### DISCUSSION TITLE
Resolving the problem of what constitute a valid Modify methods

### DISCUSSION THREAD

**Discussion by adecler** [BHoM/BHoM#1031] [Ideas]
As discussed recently within the framework of https://github.com/BHoM/admin/issues/9 and https://github.com/BHoM/admin/issues/11, some `Modify` methods are still a cause for headache due to the fact that they have to return a different type than its first input.  This is generally caused by the need for those methods to combine a `Convert` with a `Modify`. For example `Project` on a `Circle` can return an `Ellipse`.

With the need to better handle the paradigm of a mutable C# code vs an Immutable UI script, this problem becomes even more urgent to address once and for all. For the last few days, we believed we had reached a nice solution by keeping the cloning isolated in the UI and forcing ALL `Modify` methods to directly modify the input object itself. By having all `Modify` methods returning void, we would then make it very clear what their purpose is. 

Sadly, some of the current `Modify` methods are proving difficult to fit into that model. Which leads to another round of brainstorming. Personally,  I strongly believe that a solution needs to come at the level of the framework (the same way we handle cloning in the UI component instead of in each method). I don't have a ready-made solution though. Hence the need for this discussion. 

In order to help better understand the exceptions we currently have, here's the list of `Modify` methods that have an output type different than their first input type:

<details>
<summary>Full list</summary>
<p>

```
IFSContainer BH.Engine.Adapters.Filing.Modify.Rename(FSDirectory directory, String name)
IFSContainer BH.Engine.Adapters.Filing.Modify.Rename(FSFile file, String name)
Void BH.Engine.Data.Modify.Add(PointMatrix<T> matrix, Point position, T data)
Void BH.Engine.Data.Modify.Add(PriorityQueue<T> queue, T item)
Void BH.Engine.Data.Modify.AddLink(Graph<T> graph, GraphLink<T> link)
Void BH.Engine.Data.Modify.AddLink(Graph<T> graph, GraphNode<T> from, GraphNode<T> to, Double weight, Boolean bothDirection)
Void BH.Engine.Data.Modify.AddNode(Graph<T> graph, GraphNode<T> node)
GraphNode<T> BH.Engine.Data.Modify.AddNode(Graph<T> graph, T value)
T BH.Engine.Data.Modify.Peek(PriorityQueue<T> queue)
T BH.Engine.Data.Modify.Pop(PriorityQueue<T> queue)
Void BH.Engine.Data.Modify.RemoveLink(Graph<T> graph, GraphLink<T> link)
Void BH.Engine.Data.Modify.RemoveLink(Graph<T> graph, GraphNode<T> from, GraphNode<T> to, Boolean bothDirection)
Void BH.Engine.Data.Modify.RemoveNode(Graph<T> graph, GraphNode<T> node)
List<IRequest> BH.Engine.Data.Modify.Requests(ILogicalRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalAndRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalNotRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalOrRequest request)
List<T> BH.Engine.Diffing.Modify.SetHashFragment(IEnumerable<T> objs, DiffConfig diffConfig)
List<Panel> BH.Engine.Environment.Modify.Split(Panel panel, List<Line> cuttingLines)
List<Panel> BH.Engine.Environment.Modify.Split(Panel panel, Panel cuttingPanel)
List<Opening> BH.Engine.Environment.Modify.SplitOpeningByGeometry(Opening opening, List<Polyline> polylines)
List<Panel> BH.Engine.Environment.Modify.SplitPanelByGeometry(Panel panel, List<Polyline> polylines)
Polyline BH.Engine.Geometry.Modify.CollapseToPolyline(Arc curve, Double angleTolerance, Int32 maxSegmentCount)
Polyline BH.Engine.Geometry.Modify.CollapseToPolyline(Circle curve, Double angleTolerance, Int32 maxSegmentCount)
Polyline BH.Engine.Geometry.Modify.CollapseToPolyline(ICurve curve, Double angleTolerance, Int32 maxSegmentCount)
Polyline BH.Engine.Geometry.Modify.CollapseToPolyline(Line curve, Double angleTolerance, Int32 maxSegmentCount)
Polyline BH.Engine.Geometry.Modify.CollapseToPolyline(PolyCurve curve, Double angleTolerance, Int32 maxSegmentCount)
ICurve BH.Engine.Geometry.Modify.Extend(Arc curve, Double start, Double end, Boolean tangentExtensions, Double tolerance)
ICurve BH.Engine.Geometry.Modify.Project(Arc arc, Plane p)
ICurve BH.Engine.Geometry.Modify.Project(Circle circle, Plane p)
ICurve BH.Engine.Geometry.Modify.Project(Ellipse ellipse, Plane p)
ICurve BH.Engine.Geometry.Modify.ProjectAlong(Arc arc, Plane plane, Vector vector)
ICurve BH.Engine.Geometry.Modify.ProjectAlong(Circle circle, Plane plane, Vector vector)
ICurve BH.Engine.Geometry.Modify.ProjectAlong(Ellipse ellipse, Plane plane, Vector vector)
ICurve BH.Engine.Geometry.Modify.Rotate(Ellipse curve, Point origin, Vector axis, Double rad)
ICurve BH.Engine.Geometry.Modify.Scale(Arc curve, Point origin, Vector scaleVector)
ICurve BH.Engine.Geometry.Modify.Scale(Circle curve, Point origin, Vector scaleVector)
ICurve BH.Engine.Geometry.Modify.Scale(Ellipse curve, Point origin, Vector scaleVector)
ICurve BH.Engine.Geometry.Modify.SetGeometry(Arc curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(Circle curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(Ellipse curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(Line curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(NurbsCurve curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(PolyCurve curve, ICurve newCurve)
ICurve BH.Engine.Geometry.Modify.SetGeometry(Polyline curve, ICurve newCurve)
List<Arc> BH.Engine.Geometry.Modify.SplitAtPoints(Arc arc, List<Point> points, Double tolerance)
List<ICurve> BH.Engine.Geometry.Modify.SplitAtPoints(Circle circle, List<Point> points, Double tolerance)
List<ICurve> BH.Engine.Geometry.Modify.SplitAtPoints(ICurve curve, List<Point> points, Double tolerance)
List<Line> BH.Engine.Geometry.Modify.SplitAtPoints(Line line, List<Point> points, Double tolerance)
List<PolyCurve> BH.Engine.Geometry.Modify.SplitAtPoints(PolyCurve curve, List<Point> points, Double tolerance)
List<Polyline> BH.Engine.Geometry.Modify.SplitAtPoints(Polyline curve, List<Point> points, Double tolerance)
ICurve BH.Engine.Geometry.Modify.Transform(Arc curve, TransformMatrix transform)
ICurve BH.Engine.Geometry.Modify.Transform(Circle curve, TransformMatrix transform)
ICurve BH.Engine.Geometry.Modify.Transform(Ellipse curve, TransformMatrix transform)
ISurface BH.Engine.Geometry.Modify.Transform(ISolid geometry, TransformMatrix transform)
ICurve BH.Engine.Geometry.Modify.Translate(Ellipse curve, Vector transform)
Object BH.Engine.Reflection.Modify.CastGeneric(IEnumerable<T> source, Type genericType)
List<MethodInfo> BH.Engine.Reflection.Modify.SortExtensionMethods(IEnumerable<MethodInfo> methods, Type type)
Void BH.Engine.Serialiser.Modify.ApplyTaggedName(T obj, String str)
```
</p>
</details>  

To make this a bit more digestable, here's the same set of just namespace and method name:
```
BH.Engine.Adapters.Filing.Modify.Rename
BH.Engine.Data.Modify.Add
BH.Engine.Data.Modify.AddLink
BH.Engine.Data.Modify.AddNode
BH.Engine.Data.Modify.Peek
BH.Engine.Data.Modify.Pop
BH.Engine.Data.Modify.RemoveLink
BH.Engine.Data.Modify.RemoveNode
BH.Engine.Data.Modify.Requests
BH.Engine.Diffing.Modify.SetHashFragment
BH.Engine.Environment.Modify.Split
BH.Engine.Environment.Modify.SplitOpeningByGeometry
BH.Engine.Environment.Modify.SplitPanelByGeometry
BH.Engine.Geometry.Modify.CollapseToPolyline
BH.Engine.Geometry.Modify.Extend
BH.Engine.Geometry.Modify.Project
BH.Engine.Geometry.Modify.ProjectAlong
BH.Engine.Geometry.Modify.Rotate
BH.Engine.Geometry.Modify.Scale
BH.Engine.Geometry.Modify.SetGeometry
BH.Engine.Geometry.Modify.SplitAtPoints
BH.Engine.Geometry.Modify.Transform
BH.Engine.Geometry.Modify.Translate
BH.Engine.Reflection.Modify.CastGeneric
BH.Engine.Reflection.Modify.SortExtensionMethods
BH.Engine.Serialiser.Modify.ApplyTaggedName

```

@al-fisher , @alelom , @IsakNaslundBh , @FraserGreenroyd , @pawelbaran, @rolyhudson , FYI

**Comment by adecler** [comment]
Trying to categorise those methods:

### Already return void 

so ahead of anyone else 😄 

```
Void BH.Engine.Data.Modify.Add(PriorityQueue<T> queue, T item)
Void BH.Engine.Data.Modify.Add(PointMatrix<T> matrix, Point position, T data)
Void BH.Engine.Data.Modify.RemoveLink(Graph<T> graph, GraphNode<T> from, GraphNode<T> to, Boolean bothDirection)
Void BH.Engine.Data.Modify.RemoveLink(Graph<T> graph, GraphLink<T> link)
Void BH.Engine.Data.Modify.AddLink(Graph<T> graph, GraphNode<T> from, GraphNode<T> to, Double weight, Boolean bothDirection)
Void BH.Engine.Data.Modify.AddLink(Graph<T> graph, GraphLink<T> link)
Void BH.Engine.Data.Modify.AddNode(Graph<T> graph, GraphNode<T> node)
Void BH.Engine.Data.Modify.RemoveNode(Graph<T> graph, GraphNode<T> node)
Void BH.Engine.Serialiser.Modify.ApplyTaggedName(T obj, String str)

```

### Returning a List while input is an object

those don't feel like `Modify` methods to me if I am honest.

```
List<IRequest> BH.Engine.Data.Modify.Requests(ILogicalRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalAndRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalOrRequest request)
List<IRequest> BH.Engine.Data.Modify.Requests(LogicalNotRequest request)
List<Panel> BH.Engine.Environment.Modify.Split(Panel panel, List<Line> cuttingLines)
List<Panel> BH.Engine.Environment.Modify.Split(Panel panel, Panel cuttingPanel)
List<Opening> BH.Engine.Environment.Modify.SplitOpeningByGeometry(Opening opening, List<Polyline> polylines)
List<Panel> BH.Engine.Environment.Modify.SplitPanelByGeometry(Panel panel, List<Polyline> polylines)
List<Arc> BH.Engine.Geometry.Modify.SplitAtPoints(Arc arc, List<Point> points, Double tolerance)
List<ICurve> BH.Engine.Geometry.Modify.SplitAtPoints(Circle circle, List<Point> points, Double tolerance)
List<Line> BH.Engine.Geometry.Modify.SplitAtPoints(Line line, List<Point> points, Double tolerance)
List<PolyCurve> BH.Engine.Geometry.Modify.SplitAtPoints(PolyCurve curve, List<Point> points, Double tolerance)
List<Polyline> BH.Engine.Geometry.Modify.SplitAtPoints(Polyline curve, List<Point> points, Double tolerance)
List<ICurve> BH.Engine.Geometry.Modify.SplitAtPoints(ICurve curve, List<Point> points, Double tolerance)
```

### Returning a List while input is an IEnumerable

That one is pretty straightforward to fix.

```
List<MethodInfo> BH.Engine.Reflection.Modify.SortExtensionMethods(IEnumerable<MethodInfo> methods, Type type)
List<T> BH.Engine.Diffing.Modify.SetHashFragment(IEnumerable<T> objs, DiffConfig diffConfig)
```

### The Geometry methods

I have kept then in the summarized form to keep things easier to read. 

Some are debatable and clearly the best illustration of our current problem:

```
BH.Engine.Geometry.Modify.Extend
BH.Engine.Geometry.Modify.Project
BH.Engine.Geometry.Modify.ProjectAlong
BH.Engine.Geometry.Modify.Scale
BH.Engine.Geometry.Modify.Transform
```

Some just need fixing because they don't need a different input:

```
BH.Engine.Geometry.Modify.Rotate
BH.Engine.Geometry.Modify.SetGeometry
BH.Engine.Geometry.Modify.Translate
```

One feels very much like a `Compute` method to me. To reinforce that, there is already a `CollapseToScaledPolyline` method in `Compute` 

```
BH.Engine.Geometry.Modify.CollapseToPolyline
```

### The odd ones

```
Object BH.Engine.Reflection.Modify.CastGeneric(IEnumerable<T> source, Type genericType)
```

That's a `Convert`, not a `Modify`

---
```
T BH.Engine.Data.Modify.Pop(PriorityQueue<T> queue)
```

We could solve this one by using a `out` parameter

---

```
GraphNode<T> BH.Engine.Data.Modify.AddNode(Graph<T> graph, T value)
```

Not sure why the return type is different on this one to be honest

---

```
T BH.Engine.Data.Modify.Peek(PriorityQueue<T> queue)
```

This is a `Query` method, not a `Modify`

---

```
IFSContainer BH.Engine.Adapters.Filing.Modify.Rename(FSFile file, String name)
IFSContainer BH.Engine.Adapters.Filing.Modify.Rename(FSDirectory directory, String name)
```
Hum, @alelom  ?

**Comment by adecler** [comment]
So, to summarise, we are left with only five methods that are a bit tricky:

```
BH.Engine.Geometry.Modify.Extend
BH.Engine.Geometry.Modify.Project
BH.Engine.Geometry.Modify.ProjectAlong
BH.Engine.Geometry.Modify.Scale
BH.Engine.Geometry.Modify.Transform
```

Unless I missed something ? @IsakNaslundBh ?

**Reply by IsakNaslundBh** [reply]
I think those are probably the main ones, yes. As discussed, mainly for the reason of what does a the "modify" word imply. Do we look at it on the level of the more zoomed out level, as in "projecting a curve" or do we look at it at the more detailed level "project a circle".

The only other one I can think of is collections where sometimes it can be nice to have input as IEnumerbale to allow for calls to the method via linq without having to use a "ToList()" while it still might be useful to return a list. Vague inputs - distinct outputs principle, especially for the cases where it is actually the obejcts inside the collection that are being modified, rather than the collection itself. As in, a "Translate" method working on a collection of points (as a made up example) I think IEnumerable input could be ok, while a "Add" method should not be ok.

**Comment by adecler** [comment]
I think it is very important to make it clear why we are even doing this in the first place. So let me try to put a bit more clarity:

In the very early days of the BHoM, we decided that a `Modify` method would always return a new modified copy of the input object. So basically, every single `Modify` method is a `Clone` + `Modify`. This means immutability is enforced on both UI and C#. It sounded like a good plan at the time since it removes a discrepancy in approach between coding in C# and scripting in the UI.

The problem is that a lot of `Modify` methods are actually incorrect. Some don't clone the object. Some either return the original or a clone depending of the path through the if conditions. Some even return `null` while they should just return (a clone of) the original object with a message stating that the object couldn't be modify.

So this `immutable in C#` idea that sounded good at first has proven to be difficult to enforce. This translates into `Modify` methods with unpredictable behaviour. The alternative is therefore to accept the mutability of C# and enforce immutability on the UI interface. This is why, to make things very clear and easy to enforce, we decided to have the `Modify` methods to always return void. That way, we remove any risk of confusion since it becomes clear that the original object is being modified.

That obviously leaves us with the few methods that were using the `Clone` + `Modify` interface to sneak in a `Convert` in between the two.

**Comment by al-fisher** [comment]
Great - really useful. and agreed with above.

In terms of clarifying approach for the tricky cases - reading and musing on historic definitions https://github.com/BHoM/documentation/wiki/BHoM_Engine-Classes got me thinking.

The rules for _"Modify"_ returning same return type have always been clear, however, I think the impact this has had on our code organisation has been to put too much emphasis on _"Compute"_ methods.

Compute I feel has always been a bit of a catch all - a place to put methods that do not have a good home elsewhere. Indeed even our explanation in the Wiki for _"Compute"_ has two very different and discrete cases. https://github.com/BHoM/documentation/wiki/BHoM_Engine-Classes#compute

So below is a first/new attempt at more precise classification. 
It introduces a new Class, and tighter restriction on Compute. 
What do people think?

Part of the problem with Modify I feel is - often we _want to not put the code in the compute soup, as feels linguistically more logical for the method to sit in Modify. In this way the new Transform could help?

#### BHoM_Engine Classes

| Class     | Description | Extension Method | Return Type equal to Input | Output preserves meaning/values of Input |   |
|-----------|-----------|------------------|--------------------------------|----------------------|---|
| Create   | Generate new object  | No               | _N/A_ <sup>1</sup>   | _N/A_ <sup>1</sup>                |   |
| Query   | Obtain derived property  | Yes              | No                             | No                   |   |
| Convert   | Translate to new object type, with intent to preserve information/meaning | Yes              | No                             | Yes  <sup>2</sup>                   |   |
| Modify    | Manipulate an object, changing defining properties/adding fragments | Yes              | Yes                            | No                   |   |
| Transform | Manipulate an object resulting in an output of new type(s) of object | Yes              | No                             | No                   |   |
| Compute   | For computationally more intensive methods, iterative process, solvers. Algorithms that operate on objects/parameters, but are not extension methods to existing object types  | No               | _N/A_ <sup>1</sup>   | _N/A_ <sup>1</sup>                |   |


<sup>1</sup> Not applicable as method is not an extension method, so no primary input type to reference
<sup>2</sup> Note here the intent for a Convert method is for information to be persistent across translation between representations and object models. Examples like `ToJson` and `FromJson` - full fidelity can be guaranteed. To and from third party object models - can not necessarily be lossless

**Reply by al-fisher** [reply]
The new proposed Transform could be a good home for both methods like `Split`, `CollapseToPolyline`, `Project`,  and `Scale` alike.
And making the Compute class no longer accept extension methods (closer to Create class) could also help define the line.

What do people think?

**Reply by alelom** [reply]
I agree completely with @adecler and I would like to solve this problem too. 
I also like a lot @al-fisher proposal; introducing a new class of methods (Transform) is very much the only solution I can think of too. 

To recap my position: even before @adecler rightful considerations on mutability/immutability, I think that the main challenge we are facing is that _we are not distinguishing between two valid interpretations of the Modify methods_. The two being:
1. Are you modifying the `class` object (e.g. you modify a `radius` property of an ellipse)?
2. Are you modifying the _meaning_ or _concept_ that the `class` represents (i.e. are you resizing an ellipse, and perhaps by doing that are you looking to get a circle)?

For example, @IsakNaslundBh made clear that he has always referred to Modify as to (2) ("user-centric view"), while @adecler and me are thinking about (1) more ("developer clarity").

I think the the introduction of a new category is bound to solve this problem, at least partially. My main uncertainties stems from the fact that by introducing Transform _we are further separating the intention of the user from the method definition_, which instead is an advantage of having only the Modify category (@IsakNaslundBh's point I believe). Let me explain:

  a) the Modify category becomes effectively a "sub-category" of the Transform. By its definition, the Transform method has a higher "potency" that includes that of the Modify methods too. E.g. `Scale` transforms a circle by modifying its radius.  
I think that the answer to this is that we have the "`Return Type equal to Input`: `No`" requirement for the Transform category in order to fully distinguish it from the Modify.

  b) As a consequence of (a), we are perhaps complicating the user experience and development effort. The more categories we have, the more "rational" the thought process for finding/positioning methods must be.  
As a user, does "Return Type equal to Input: No" means that I need to think about all the potential outputs of a method in order to imagine whether that's located in Modify or in Transform? E.g. I want to change the size of an Ellipse, that as a consequence might become a Circle: do I look into Modify or in Transform? For circles/ellipses it's easy, but is it for all cases?  
Maybe an answer to this could be: we could have both.

I wrote the following as a thought experiment related to (b). Consider this theoretical scenario: 
- we only have a `Circle` class, and no `Ellipse` class exists in BHoM.
- we define a method `Scale` that takes a `Circle` and changes its radius. At this stage, the right location for the method is the Modify category and it should be put there.
- we introduce the `Ellipse` class. 

Do we move the code for `Scale(Circle)` from the Modify to the Transform? Or do we simply introduce another method in Transform?  
I.e. By the simple addition of a new class `Ellipse`, we need to consider that the method Scale should probably be rewritten to take an Ellipse, as it can return an Ellipse or a Circle.

This is to say that by introducing the class Transform we might have to consider that the Engine has to be checked for compliance against the existing set of BHoM classes. This check might not be trivial and automated in all cases I think.

**Reply by pawelbaran** [reply]
I agree with @alelom - comparing (1) more ("developer clarity") with (2) ("user-centric view") is a very valid diagnosis. And here I am on @IsakNaslundBh's side, so "user-centric view". We need to remember that most users will see no difference between `Modify` and `Transform` (and _will not want to see it_). On top of that, `Transform` could quickly become too small for its scope (even now, one could argue whether curve extension is a transformation).

Besides that, everything else mentioned by @alelom also makes perfect sense.

I am aware that my comment does not bring us any closer to the solution, but adding a sixth, slightly artificial class does not seem to be the solution to me (but still can be best that we can get though).

**Reply by IsakNaslundBh** [reply]
Would agree with @alelom s statements above here as well. (also replied earlier up before I had read all of this! )

Not sure adding another Engine classification will solve this for us unfortunately. Fear that would just push the problem somewhere else.

To me, the first thing we need to agree on are what do we mean with Modify, between @alelom s two options above. Think that has been the main reason for confusion and disagreement to date regarding this subject!

Might be worth a workshop/meeting regarding it quite soon to bottom it out? Based on what we agree then quite a few things will at least become a bit clearer.

Some initial pros and cons I can see:

>Are you modifying the class object (e.g. you modify a radius property of an ellipse)?

+Easy to check for
+Allows returning void to even more simply enforce the rules we set up
-Especially geometry methods can get forced into compute which might make things harder to find
-Will put more stringency on modify methods working on a collection of elements for efficiency reasons

>Are you modifying the meaning or concept that the class represents (i.e. are you resizing an ellipse, and perhaps by doing that are you looking to get a circle)?

+Makes it possible for us to put methods where we think users will expect them to be. User centric code structure
-Harder to automatically check leading to all emphasis being put on the reviewer
-void return to possible to enforce

As @adecler has pointed out above as well, the methods impacted by whatever choice we make between the two above will not impact most methods. It is mainly geometry being affected, when it sometimes need to change type during the modification. Ofc an object owning a geometry would not be affected this way, as it would just be one of its properties being replaced, still keeping the top level object the same.

This, and collection based methods. I would argue it would be nice to be able to accept an IEnumerbale as an input but return a List, when modifying the obejcts inside the collection, rather than the collection itself. This is a more minor point though.

Sorry, this became longer than intended to say "I agree with @alelom and @pawelbaran "

**Comment by adecler** [comment]
Thanks guys, great comments !

Personally, I feel that the key to solve this is `Scale`. Funnily enough, when I started this discussion, `Scale` was the only one I had miscategorised in the same section as `Rotate` and `Translate`. I had completely forgotten about non-uniform scaling of a circle. While I have no problem with `Project` being in `Compute`, the only category that makes sense for `Scale` is `Modify`. Especially if we consider every single other case of scaling. When I realised that, it was already late. So I did the only reasonable thing to do: I updated my comments here and turned off my computer. 

Looking at it with fresh eyes, it still feels like there is no great categorisation solution to this. So allow me to throw a curve ball just to force us to see things from a different perspective. 

I think that the reason why all our tricky cases are in `Geometry` might be because this is the only namespace (that I can think of) that has multiple representations of the same thing. An ellipse and an arc are exceptional cases of a curve. A circle is an exceptional case of an ellipse as well as an exceptional case of an arc. The `Circle` type is there purely for convenience but we could replace it with an `Ellipse` in every single case. Pushing that idea to the extreme, we could use NURBS curves everywhere we use any other form of curve and there would be no loss of information. So what would I change if `NurbsCurve` was my only class of curve? Probably just this (and keep `Polyline` as a separate class ?):

```c#
public class NurbsCurve : ICurve
{
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        public virtual List<Point> ControlPoints { get; set; } = new List<Point>();

        public virtual List<double> Weights { get; set; } = new List<double>();

        public virtual List<double> Knots { get; set; } = new List<double>();

        public virtual CurveType Type { get; set; } = CurveType.NurbsCurve;
}

public enum CurveType
{
        NurbsCurve,
        Polyline,
        Line,
        Arc,
        Ellipse,
        Circle
}
```

Then use the `CurveType` to handle things like conversion to and from external libraries, text display and methods like `Query.Radius` that would only make sense for some of the types. 

I am not saying we should do this (or should we ? 😉 ) but I think that, to solve this problem, we need to have a good understanding of its causes. And this might be one pointer.

**Comment by adecler** [comment]
Alternative idea:

Allowing Modify methods to have two options for the return type would help inform the code user what to expect:
- return `void`: the object is modified directly
- returning a type: the object returned is a new one

**Comment by al-fisher** [comment]
Thanks all - great conversations. As discussed on call with @adecler @alelom @IsakNaslundBh _et al._ have updated summary table with new columns.

_To summarise:_ 
- Requirement for new sub categories of the modify class based on the return type of the method.
  - return void: the object is modified directly
  - returning a type: the object returned is a new one
- Compute methods are not extension methods


| Class     | Description | Extension Method |Returns Void  | Guaranteed to not modify input | Output preserves meaning & values of Input | 
|-----------|-----------|------------------|--------------------------------|----------------------|---|
| **Create**   | Generate new object  | No               | No  | **Yes**  | _N/A_ <sup>1</sup>                | 
| **Query**   | Obtain derived property.   | **Yes**              | No                             | **Yes**  |No                   | 
| **Convert**   | Translate to new object type, with intent to preserve information/meaning | **Yes**              | No                             | **Yes** | **Yes**  <sup>2</sup>                   |  
| **Modify**    | Manipulate an object **directly**. _i.e. changing defining properties/adding fragments_ | **Yes**              | **Yes** | No                   | No  |
| **Modify** | Manipulate an object **returning new object/type of object(s)** | **Yes**              | No                             | **Yes** | No                   |  
| **Compute**   | For computationally more intensive methods, iterative process, solvers. Algorithms that operate on objects/parameters, but are not extension methods to existing object types  | No               | No  | **Yes** | _N/A_ <sup>1</sup>                |  


<sup>1</sup> Not applicable as method is not an extension method, so no primary input type to reference
<sup>2</sup> Note here the intent for a Convert method is for information to be persistent across translation between representations and object models. Examples like `ToJson` and `FromJson` - full fidelity can be guaranteed. To and from third party object models - can not necessarily be lossless

**Reply by al-fisher** [reply]
If happy with above (please do confirm/add any comments for further tweaks) - suggest we then plan to:
1.  Update Compliance guidelines and Wiki etc. 
2. Ensure the Framework supports both the new rules, and also versioning of legacy code as mentioned on call. Any issues here to be raised, if not already, and closed in 4.0
3. Then finally, also start raising some issues across engines to update methods as needed - planned for close out in 4.1. After completion of the above
